enhancements/authentication/allow-users-to-manage-their-own-tokens.md

```diff
@@ -0,0 +1,153 @@
+how the self-management of users' access tokens can be improved.
+
+## Motivation
+
+It is quite easy for users to retrieve tokens but it is also not so hard
+for them to be able to lose access to them - people don't usually search
+their cookies for the value of the token, or don't keep history of the
+tokens they retrieve by issuing the `oc login` command multiple times.
+
+### Goals
+
+Allow users to:
+- list their own tokens
+- view details of a specific token
+- delete their own tokens
+
+### Non-Goals
+
+- enable handling sessions of services that use the OpenShift access tokens
+
+## Proposal
+
+### User Stories [optional]
+
+Detail the things that people will be able to do if this is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+
+#### Story 1
+A user is able to see all of the tokens that were issued for them, including
+the name of the OAuth2 client that requested it and the scopes of it, making
+it easier for them to audit their accesses.
+
+#### Story 2
+A user gets a friendlier way to remove their own tokens,
+`oc delete oauthaccesstoken <sha256-tokenhash>` uses generic deletion code
+which attempts to `list` the token after its deletion, and that ends up
+with a `forbidden` message since the user is usually not allowed to `list`
+their tokens.
+
+### Implementation Details/Notes/Constraints [optional]
+
+The OAuth API server gets a new API endpoint - `personaltokens`. This endpoint
+allows to `get`, `list` and `delete` tokens that belong to the current user. The
+user is determined based on the context of the request, as supplied by the generic
+API server logic.
+
+Upon receiving a REST request to the new endpoint, the API server internally
+wires get/list/delete requests to the `oauthaccesstoken` API endpoint.
+
+To list the tokens internally, a field selector of the form `userName=<userName>`
+is used. The `userName` field selector must not be overridden by any other selector.
+
+When getting or deleting a token, there must be a strict match of `token.userName`
+and the `userName` retrieved as described above. If such a match does not occur,
+the API server must return "404 - Not Found" so that it is not possible to guess
+other people's token object names.
+
+### Risks and Mitigations
+
+The proposal is risky as it allows (limited) access to a resource that was
+previously considered confidential, and that serves as means to authenticate
+to OpenShift.
+
+The level of confidentiality of the `OAuthAccessToken` objects was lowered in
+https://github.com/openshift/enhancements/pull/323 and therefore, should a
+token leak, it is no longer possible to use the name of such an object to
+log in as a different user. On the other hand, the token still contains
+information about the user and which services they might be accessing.
```

Comment:
was this different before? Everybody could either GET (see my quesion above) or at least DELETE tokens. So you had a way to access the token object content even before.

Reply:
I don't think we allow GET on oauthaccesstokens. But I tried the DELETE of a low privileged user on kube:admin and was surprised for this to work. I did not know that. This is weird.